### PR TITLE
Feature/generate pattern docs merge

### PR DESCRIPTION
**Pull request by loparcog**
Merged with master, generating docs seems to work without issue

### CODE REVIEW

File: auxiliary/scripts/pull-descriptions.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            fname = ("docs/patterns/%s/%s.md" % (file_dict['name'].lower(), class_dict['classname'].lower()))
+            f = open(fname, "w")
+            f.write("\n\n".join(class_content))
+            f.close()
+
+def parse_class_data(args):
+    # Get all pattern files, excluding __init__'s
+    pattern_files = list(sorted(glob.glob("isobar/pattern/[!_]*.py", recursive=True)))
+
+    allDicts = []
```
Comment (ideoforms) [inline auxiliary/scripts/pull-descriptions.py:78]: To stick to Python [PEP8](https://peps.python.org/pep-0008/) guidelines, avoid camelCase and use underscored snake_case for variable names etc.

File: auxiliary/scripts/pull-descriptions.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    for fname in pattern_files:
+        basename = os.path.basename(fname)
+
+        # Make a new dict for this file
+        fileDict = {
+            "name": basename[:-3],
```
Comment (ideoforms) [inline auxiliary/scripts/pull-descriptions.py:85]: Better to use `os.path.splitext` to separate prefix/suffix

File: auxiliary/scripts/pull-descriptions.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            "classes": []
+        }
+
+        contents = open(fname, "r").read()
+
+        classregex = 'class\s[\w():\s]+"""((?!""").)*"""'
```
Comment (ideoforms) [inline auxiliary/scripts/pull-descriptions.py:91]: Impressive regex-fu! 🥷🏻

File: auxiliary/scripts/pull-descriptions.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                        # Format a code block around the name of the arg
+                        l = "`%s" % l.strip()
+                        l = re.sub(":", "`:", l, 1)
+                        arguments[i] = l
+                        i += 1
+                    print(arguments)
+                    arguments = "\n\n".join(arguments)
+
+            # Gather all information into dict
+            classDict = {
```
Comment (ideoforms) [inline auxiliary/scripts/pull-descriptions.py:141]: Same comment as above re camelCase - `class_dict` preferred here

File: auxiliary/scripts/pull-descriptions.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        contents += "\n\n"
+    # Output to library.md
+    f = open("docs/patterns/library.md", "w")
+    f.write(contents)
+    f.close()
+
+def generate_class_pages(class_data):
+    for file_dict in class_data:
+        # Check if folder exists
+        if not os.path.exists("docs/patterns/%s" % file_dict['name'].lower()):
```
Comment (ideoforms) [inline auxiliary/scripts/pull-descriptions.py:51]: Better in general to use `os.path.join` to combine path elements - create a `root_dir` variable containing `docs/patterns`, and then use `os.path.join` to create pathnames under this.

File: auxiliary/scripts/pull-descriptions.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        contents += "| Class | Function |\n"
+        contents += "|-|-|\n"
+        for class_dict in file_dict["classes"]:
+            contents += "| [%s](%s/%s.md) | %s |\n" % (class_dict["classname"],
+                                                       file_dict["name"],
+                                                       class_dict["classname"].lower(),
+                                                       class_dict["short_description"])
+        contents += "\n\n"
+    # Output to library.md
+    f = open("docs/patterns/library.md", "w")
```
Comment (ideoforms) [inline auxiliary/scripts/pull-descriptions.py:44]: In modern Python, there's a nice way to do this using the `with` keyword:

```
with open(pathname, "w") as fd:
   f.write(contents)
```

As soon as the `with` scope is exited, the fd is closed automatically.

**Review by ideoforms — CHANGES_REQUESTED** [review]
Awesome PR - thanks for all your work on this! The output is fantastic and will make reading the documentation a much nicer experience. Great code quality and clarity too. 

I think it would be great to integrate your new logic into the main `generate-docs` script, so we would have simply:

```
generate-docs.py --library # generates the pattern library including markdown index
generate-docs.py --readme # generates the readme
generate-docs.py --all # generates both
```

Feel free to restructure/refactor/throw away elements of my old generate-docs code. 

It would also be great to have an index for each pattern category - auto-generating an index page for (e.g) `/patterns/chance/index.md` that contains an index of each class within each category, just a list of bulletpoints with its one-line description and a link to the pattern documentation page.

Added a few comments inline, but it's mostly fairly cosmetic stuff and a few tips.

**Comment by loparcog** [comment]
Going to get on the above changes over the weekend and I'll throw a progress update in here on Sunday

File: auxiliary/scripts/pull-descriptions.py (hunk 1 above)
Comment (loparcog) [reply auxiliary/scripts/pull-descriptions.py]: Done for all camelCase variables, also added onto some other two-word  vars that would be helped with the readability and consistency

File: auxiliary/scripts/pull-descriptions.py (hunk 2 above)
Comment (loparcog) [reply auxiliary/scripts/pull-descriptions.py]: Done in newest update

File: auxiliary/scripts/pull-descriptions.py (hunk 4 above)
Comment (loparcog) [reply auxiliary/scripts/pull-descriptions.py]: Also done

File: auxiliary/scripts/pull-descriptions.py (hunk 5 above)
Comment (loparcog) [reply auxiliary/scripts/pull-descriptions.py]: Did for this and two other instances where we were combining `docs/patterns` with some variable. There are other paths still lingering (ie. `with open("docs/patterns/library.md", "w") as f:`) but I imagine there isn't much benefit to swapping that out with `os.path.join`, right?

File: auxiliary/scripts/pull-descriptions.py (hunk 6 above)
Comment (loparcog) [reply auxiliary/scripts/pull-descriptions.py]: Done for all file open instances

**Comment by loparcog** [comment]
I'm getting the index files going now, but I wanted to confirm for the `generate-docs.py` args, did you want these to just print the output like how `generate-docs.py` is doing currently or actually write the markdown files like how the `--generate-markdown` flag is doing?

**Comment by ideoforms** [comment]
@loparcog Let's switch to writing the files directly, much simpler usage that way. Might be a bit more regex-fu needed to overwrite only the relevant part of the README!

**Comment by loparcog** [comment]
Newest push has a pretty big load of things added in
- Index pages are now a thing and are referenced in the main pattern classes library page by clicking on the name of the file
- README updater has been put in place in `generate-docs.py`, the only thing it changes right now is it lists the files in alphabetical order for the `Pattern Classes` section, which can easily be changed just up to preference
- Arguments have been added into `generate-docs.py` (as well as all of the code from `pull-descriptions.py`) to tuck everything in and have it working with the arguments you provided. If you want to keep the original code in `pull-descriptions` and just reference it instead lemme know

**Review by ideoforms — APPROVED** [review]

**Pull request merged**